File: 31 Depth Charge/javascript/depthcharge.js
// DEPTH CHARGE
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(30) + "DEPTH CHARGE\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("DIMENSION OF THE SEARCH AREA");
    g = Math.floor(await input());
    n = Math.floor(Math.log(g) / Math.log(2)) + 1;
    print("YOU ARE THE CAPTAIN OF THE DESTROYER USS COMPUTER\n");
    print("AN ENEMY SUB HAS BEEN CAUSING YOU TROUBLE.  YOUR\n");
    print("MISSION IS TO DESTROY IT.  YOU HAVE " + n + " SHOTS.\n");
    print("SPECIFY DEPTH CHARGE EXPLOSION POINT WITH A\n");
    print("TRIO OF NUMBERS -- THE FIRST TWO ARE THE\n");
    print("SURFACE COORDINATES; THE THIRD IS THE DEPTH.\n");
    do {
        print("\n");
        print("GOOD LUCK !\n");
        print("\n");
        a = Math.floor(Math.random() * g);
        b = Math.floor(Math.random() * g);
        c = Math.floor(Math.random() * g);
        for (d = 1; d <= n; d++) {
            print("\n");
            print("TRIAL #" + d + " ");
            str = await input();
            x = parseInt(str);
            y = parseInt(str.substr(str.indexOf(",") + 1));
            z = parseInt(str.substr(str.lastIndexOf(",") + 1));
            if (Math.abs(x - a) + Math.abs(y - b) + Math.abs(z - c) == 0)
                break;
            if (y > b)
                print("NORTH");
            if (y < b)
                print("SOUTH");
            if (x > a)
                print("EAST");
            if (x < a)
                print("WEST");
            if (y != b || x != a)
                print(" AND");
            if (z > c)
                print(" TOO LOW.\n");
            if (z < c)
                print(" TOO HIGH.\n");
            if (z == c)
                print(" DEPTH OK.\n");
            print("\n");
        }
        if (d <= n) {
            print("\n");
            print("B O O M ! ! YOU FOUND IT IN " + d + " TRIES!\n");
        } else {
            print("\n");
            print("YOU HAVE BEEN TORPEDOED!  ABANDON SHIP!\n");
            print("THE SUBMARINE WAS AT " + a + "," + b + "," + c + "\n");
        }
        print("\n");
        print("\n");
        print("ANOTHER GAME (Y OR N)");
        str = await input();
    } while (str.substr(0, 1) == "Y") ;
    print("OK.  HOPE YOU ENJOYED YOURSELF.\n");
}

main();

File: 31 Depth Charge/java/DepthCharge.java
import java.util.Scanner;
import java.lang.Math;

/**
 * Game of Depth Charge
 * <p>
 * Based on the BASIC game of Depth Charge here
 * https://github.com/coding-horror/basic-computer-games/blob/main/31%20Depth%20Charge/depthcharge.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's BASIC game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 * 
 * Converted from BASIC to Java by Darren Cardenas.
 */
 
public class DepthCharge {  
  
  private final Scanner scan;  // For user input
  
  public DepthCharge() {
    
    scan = new Scanner(System.in);
    
  }  // End of constructor DepthCharge 

  public void play() {
    
    showIntro();
    startGame();
    
  }  // End of method play  
    
  private static void showIntro() {
    
    System.out.println(" ".repeat(29) + "DEPTH CHARGE");
    System.out.println(" ".repeat(14) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
    System.out.println("\n\n");

  }  // End of method showIntro        
  
  private void startGame() {

    int searchArea = 0;
    int shotNum = 0;
    int shotTotal = 0;    
    int shotX = 0;
    int shotY = 0;
    int shotZ = 0;
    int targetX = 0;
    int targetY = 0;
    int targetZ = 0;
    int tries = 0;
    String[] userCoordinates;
    String userResponse = "";
    
    System.out.print("DIMENSION OF SEARCH AREA? ");
    searchArea = Integer.parseInt(scan.nextLine());
    System.out.println("");
    
    shotTotal = (int) (Math.log10(searchArea) / Math.log10(2)) + 1;      
    
    System.out.println("YOU ARE THE CAPTAIN OF THE DESTROYER USS COMPUTER");
    System.out.println("AN ENEMY SUB HAS BEEN CAUSING YOU TROUBLE.  YOUR");
    System.out.println("MISSION IS TO DESTROY IT.  YOU HAVE " + shotTotal + " SHOTS.");
    System.out.println("SPECIFY DEPTH CHARGE EXPLOSION POINT WITH A");
    System.out.println("TRIO OF NUMBERS -- THE FIRST TWO ARE THE");
    System.out.println("SURFACE COORDINATES; THE THIRD IS THE DEPTH.");
    
    // Begin outer while loop    
    while (true) {         
    
      System.out.println("");
      System.out.println("GOOD LUCK !");
      System.out.println("");
      
      targetX = (int) ((searchArea + 1) * Math.random());
      targetY = (int) ((searchArea + 1) * Math.random());
      targetZ = (int) ((searchArea + 1) * Math.random());
      
      // Begin loop through all shots
      for (shotNum = 1; shotNum <= shotTotal; shotNum++) {
      
        // Get user input
        System.out.println("");
        System.out.print("TRIAL # " + shotNum + "? ");        
        userResponse = scan.nextLine();
        
        // Split on commas
        userCoordinates = userResponse.split(",");
        
        // Assign to integer variables
        shotX = Integer.parseInt(userCoordinates[0].trim());
        shotY = Integer.parseInt(userCoordinates[1].trim());
        shotZ = Integer.parseInt(userCoordinates[2].trim());       
      
        // Win condition
        if (Math.abs(shotX - targetX) + Math.abs(shotY - targetY) 
            + Math.abs(shotZ - targetZ) == 0) {

          System.out.println("B O O M ! ! YOU FOUND IT IN" + shotNum + " TRIES!");
          break;
          
        }
        
        this.getReport(targetX, targetY, targetZ, shotX, shotY, shotZ);
        
        System.out.println("");                
      
      }  // End loop through all shots
      
      if (shotNum > shotTotal) {
        
        System.out.println("");
        System.out.println("YOU HAVE BEEN TORPEDOED!  ABANDON SHIP!");
        System.out.println("THE SUBMARINE WAS AT " + targetX + "," + targetY + "," + targetZ);
      }
      
      System.out.println("");
      System.out.println("");
      System.out.print("ANOTHER GAME (Y OR N)? ");               
      userResponse = scan.nextLine();      
      
      if (!userResponse.toUpperCase().equals("Y")) {
        System.out.print("OK.  HOPE YOU ENJOYED YOURSELF."); 
        return;
      }
    
    }  // End outer while loop
    
  }  // End of method startGame  
    
  public void getReport(int a, int b, int c, int x, int y, int z) {
    
    System.out.print("SONAR REPORTS SHOT WAS ");
    
    // Handle y coordinate
    if (y > b) {
      
      System.out.print("NORTH");      
      
    } else if (y < b) {
      
      System.out.print("SOUTH");
    }
    
    // Handle x coordinate
    if (x > a) {
      
      System.out.print("EAST");
      
    } else if (x < a) {
      
      System.out.print("WEST");      
    }
    
    if ((y != b) || (x != a)) {
      
      System.out.print(" AND");
    }
    
    // Handle depth
    if (z > c) {
      
      System.out.println(" TOO LOW.");
      
    } else  if (z < c) {
      
      System.out.println(" TOO HIGH.");
      
    } else {
      
      System.out.println(" DEPTH OK.");      
    }
    
    return;
    
  }  // End of method getReport
  
  public static void main(String[] args) {
    
    DepthCharge game = new DepthCharge();
    game.play();
    
  }  // End of method main
    
}  // End of class DepthCharge
